README.md
# About this Visual Studio Code extension

This extension is for people that want a more "in the face" approach on highlighting content between symbols. \
Symbols can be custom defined, the extension is not limited to only brackets. Sadly the extension name cannot be changed retroactively. \
It provides a very configurable CSS style markup for the whole text between opening and closing symbols or only the symbols themselves. \
There are preset options for all kinds of brackets.

# Features

- Highlights pairs of symbols that are at the cursor or if the cursor is in the scope of the pair
- Symbols can be configured and support regular expressions
- Multiple cursors support
- Hotkey actions to jump to the highlight pairs
- Hotkey action to select all text between a highlight pair

# Upgrade to version 3.0.0
- Please refer to the chapter [quick setup](#quick-setup-of-old-parameters) if you want to quickly get started with basic symbol support
- Please refer to the chapter [custom symbols](#custom-symbols) for more information how the new syntax for highlight symbols looks like
- The extension will try to use symbols with the old syntax, but be aware that it is deprecated and might not work as you expect it to
- Some settings were removed, which means there are no more symbols available via checkboxes => All symbols must now be configured within custom symbols. You can find a minimal example in the [minimal example](#minimal-custom-symbol-example) chapter.


# Decoration examples

| Example                            | Description                               |
| ---------------------------------- | ----------------------------------------- |
|![](assets/backgroundColor.png)     | Change background color                   |
|![](assets/blur.png)                | Blur everything not in scope              |
|![](assets/differentSettings.png)   | Highlight symbols and content differently |
|![](assets/highlightEverything.png) | Highlight symbols and content the same    |
|![](assets/highlightScope.png)      | Highlight only the symbols                |
|![](assets/multiCursor.png)         | Multiple cursor support                   |
|![](assets/textAlteration.png)      | Change the way the text looks             |

- **... Lots of additional possibilities. Just configure it the way you like it.**

**Hotkey actions:**

- Jump to/out of highlighted opening and closing symbols  
- Jump to the corresponding symbol pair
- Select the text between highlighted symbols


# Extension Settings

This extension contributes the following settings:

| Setting                                               | Description                                                                                                                                       |
| ----------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| `BracketHighlighter.enableExtension`                  | Enables the extension when set.                                                                                                                   |
| `BracketHighlighter.maxLineSearchCount`               | The maximum amount of lines to search for a matching symbol. If nothing is found by this number, nothing will be highlighted.                     |
| `BracketHighlighter.highlightScopeFromText`           | Allows highlighting when clicking inside of a scope. It will search for the scope of all enabled symbols.                                         |
| `BracketHighlighter.allowedLanguageIds`               | IDs which this extension will work on. Leaving this blank will enable it globally. Identifiers have to be separated by a comma. E.g. c,cpp,java.  |
| `BracketHighlighter.fontWeight`                       | FontWeight in CSS style. E.g. 'bold'. If separate highlighting for symbols and content is used, this parameter is used for the content.           |
| `BracketHighlighter.fontStyle`                        | FontStyle in CSS style. E.g. 'oblique'. If separate highlighting for symbols and content is used, this parameter is used for the content.         |
| `BracketHighlighter.letterSpacing`                    | LetterSpacing in CSS style. E.g. '1px'. If separate highlighting for symbols and content is used, this parameter is used for the content.         |
| `BracketHighlighter.outline`                          | Outline in CSS style. E.g. '2px dashed blue'. If separate highlighting for symbols and content is used, this parameter is used for the content.   |
| `BracketHighlighter.border`                           | Border in CSS style. E.g. '4px dotted blue'. If separate highlighting for symbols and content is used, this parameter is used for the content.    |
| `BracketHighlighter.backgroundColor`                  | BackgroundColor in CSS style. E.g. 'coral'. If separate highlighting for symbols and content is used, this parameter is used for the content.     |
| `BracketHighlighter.textColor`                        | Sets the color of the highlighted text.                                                                                                           |
| `BracketHighlighter.differentSymbolHighlightingUsed`  | If enabled, separate text decorations can be used for symbols and content.                                                                        |
| `BracketHighlighter.fontWeightSymbol`                 | Parameter is used for symbols only if separate highlighting for symbols and content is enabled: fontWeight in CSS style. E.g. 'bold'.             |
| `BracketHighlighter.fontStyleSymbol`                  | Parameter is used for symbols only if separate highlighting for symbols and content is enabled: fontStyle in CSS style. E.g. 'oblique'.           |
| `BracketHighlighter.letterSpacingSymbol`              | Parameter is used for symbols only if separate highlighting for symbols and content is enabled: letterSpacing in CSS style. E.g. '1px'.           |
| `BracketHighlighter.outlineSymbol`                    | Parameter is used for symbols only if separate highlighting for symbols and content is enabled: outline in CSS style. E.g. '2px dashed blue'.     |
| `BracketHighlighter.borderSymbol`                     | Parameter is used for symbols only if separate highlighting for symbols and content is enabled: border in CSS style. E.g. '4px dotted blue'.      |
| `BracketHighlighter.backgroundColorSymbol`            | Parameter is used for symbols only if separate highlighting for symbols and content is enabled: backgroundColor in CSS style. E.g. 'coral'.       |
| `BracketHighlighter.textDecorationSymbol`             | Parameter is used for symbols only if separate highlighting for symbols and content is enabled: textDecoration in CSS style. E.g. 'underline'.    |
| `BracketHighlighter.textColorSymbol`                  | Parameter is used for symbols only if separate highlighting for symbols and content is enabled: Sets the color of the highlighted text.           |
| `BracketHighlighter.activeInDebugMode`                | Enables the extension when debugging.                                                                                                             |
| `BracketHighlighter.blurOutOfScopeText`               | Enables a blur effect on non-highlighted text. Opacity depends on blurOpacity value.                                                              |
| `BracketHighlighter.blurOpacity`                      | Sets the opacity of the blurred out text. E.g. 0.5 for 50% visibility.                                                                            |
| `BracketHighlighter.ignoreContent`                    | Ignores content in the scope and instead only highlights the enclosing symbols.                                                                   |
| `BracketHighlighter.timeOutValue`                     | Sets a value in milliseconds how often highlighting can be triggered. A higher value will increase performance when writing, however highlighting may be delayed in some cases. Setting this to 0 will make the extension behave as it did before this option existed.                                                                                                       |
| `BracketHighlighter.customSymbols`                    | User defined symbols, which are explained in more detail in the extension description.                                                            |

Refer to https://www.w3schools.com/cssref/ for all CSS options. \
Refer to https://code.visualstudio.com/docs/languages/identifiers for available language identifiers. \
Refer to https://www.w3schools.com/cssref/css_colors_legal.asp for possible color values. \


# Hotkeys provided by extension

This extension provides several hotkeys to work with highlighted symbols and ranges.

| Hotkey setting                                            | Description                                                                                        |
| --------------------------------------------------------- | -------------------------------------------------------------------------------------------------- |
| `BracketHighlighter.toggleExtensionStatus`                | Enables/Disables the extension. (Default hotkey: Ctrl + Alt + L)                                   |
| `BracketHighlighter.jumpOutOfClosingSymbol`               | Jumps to the outside of the closing symbol. (Default hotkey: Ctrl + Alt + DownArrow)               |
| `BracketHighlighter.jumpOutOfOpeningSymbol`               | Jumps to the outside of the opening symbol. (Default hotkey: Ctrl + Alt + UpArrow)                 |
| `BracketHighlighter.jumpToClosingSymbol`                  | Jumps to the inside of the closing symbol. (Default hotkey: Ctrl + Alt + RightArrow)               |
| `BracketHighlighter.jumpToOpeningSymbol`                  | Jumps to the inside of the opening symbol. (Default hotkey: Ctrl + Alt + LeftArrow)                |
| `BracketHighlighter.selectTextInSymbols`                  | Selects the whole text between (and not including) the symbols. (Default hotkey: Ctrl + Alt + K)   |
| `BracketHighlighter.jumpBetweenOpeningAndClosingSymbols`  | Jumps between the opening and closing symbols. (Default hotkey: Ctrl + Alt + \\)                   |


# Quick setup of old parameters
If you just quickly want to quickly get a setup for all available types of brackets, simply copy the following into your ```settings.json``` file. \
For more details about the structure refer to the [custom symbols](#custom-symbols) chapter.

```json 
    "BracketHighlighter.customSymbols": [
        {
            "highlightPair": [
                {
                    "symbol": "(",
                    "isRegex": false,
                    "canBeSubstring": true
                },
                {
                    "symbol": ")",
                    "isRegex": false,
                    "canBeSubstring": true
                }
            ]
        },
        {
            "highlightPair": [
                {
                    "symbol": "[",
                    "isRegex": false,
                    "canBeSubstring": true
                },
                {
                    "symbol": "]",
                    "isRegex": false,
                    "canBeSubstring": true
                }
            ]
        },
        {
            "highlightPair": [
                {
                    "symbol": "{",
                    "isRegex": false,
                    "canBeSubstring": true
                },
                {
                    "symbol": "}",
                    "isRegex": false,
                    "canBeSubstring": true
                }
            ]
        },
        {
            "highlightPair": [
                {
                    "symbol": "<",
                    "isRegex": false,
                    "canBeSubstring": true
                },
                {
                    "symbol": ">",
                    "isRegex": false,
                    "canBeSubstring": true
                }
            ]
        }
    ]
```

# Custom symbols

The following example highlights an advanced usage scenario with regular expressions. \
Here you can see the full configuration set. It is not necessary to set everything here, for a minimal example refer to chapter [minimal custom example](#minimal-custom-symbol-example). \
The following parameters can be configured for each highlight pair:
- ```symbol: ``` Symbol to use for the pair
- ```isRegex: ``` Whether the given symbol is a regular expression
- ```canBeSubstring: ``` Whether the given symbol can be part of a substring or not. Not used if ```isRegex: true```.

---
**NOTE**

Please note the following points:
- The order of the highlight pair actually matters. The first entry of the highlight pair will be the opening symbol, the second one the closing symbol.
- Multiple symbols with the same starting or closing symbols can cause weird or unexpected behavior. Consider using a regular expression, which combines the symbols instead.
---


```json
    "BracketHighlighter.customSymbols": [
        {
            "highlightPair": [
              {
                  "symbol": "\\b(do\\b(?!:)|fn\\b)",
                  "isRegex": true,
                  "canBeSubstring": true
              },
              {
                  "symbol": "\\bend\\b",
                  "isRegex": true,
                  "canBeSubstring": true
              },
            ],
            "jumpBetweenStrategy": "toSymbolOppositeSide"
        }
    ]
```

# Minimal custom symbol example

The smalles configuration you can have with the new syntax is the following:

```json
    "BracketHighlighter.customSymbols": [
        {
            "highlightPair": [
              {
                  "symbol": "/*",
              },
              {
                  "symbol": "*/",
              },
            ],
        }
    ]
```

# Planned improvements
- This extension is feature complete. If you think something is missing please refer to the next paragraph.

# Feature requests and bug reports
Please mail them to me at dev@durzn.com or create an open issue at https://github.com/Durzn/BracketHighlighter
Thanks to everyone reporting issues and requesting new features. The extension wouldn't be anywhere near where it currently is without these.

# Special thanks
## Art_of_bini 
for the great extension icon. See more from her on her instagram page https://www.instagram.com/art_of_bini/
## You
for using this extension :)

# Support
I'm working on projects like this extension in my free time. 
If you want to buy me a coffee to keep me awake for more work on my projects, I'd greatly appreciate it.

<a href='https://ko-fi.com/H2H4Q3C6N' target='_blank'><img height='36' style='border:0px;height:36px;' src='https://storage.ko-fi.com/cdn/kofi2.png?v=3' border='0' alt='Buy Me a Coffee at ko-fi.com' /></a>

I also gladly accept ``Ada`` over this address: ``addr1qyz4hp9a5m844c5dn7nz56vget0lcx2lxg9eg02x9nyg82m7306tyvjqygl08grv79tm2jw0sh9lg8nr5u0g0qvf2nlsw376z4``

CHANGELOG.md
# Change Log

## [3.0.3]
  Reworked README

## [3.0.2]
  Fixed issue where symbols could not be right next to each other when they have a length of 1

## [3.0.1]
  Fixed issue where the correct scope could not be found under certain conditions

## [3.0.0]
  Major update with breaking changes to the configuration. \
  It is now possible to configure highlight pairs with more fine-grained control. Symbols can be a regular expression and symbols have a switch, which enables them to be substrings in the search.

## [2.4.1]
  Fixed performance issue

## [2.4.0]
  Added new hotkey action to jump between opening and closing symbols.  
  Special thanks to [@Kotsuha](https://github.com/Kotsuha) for implementing this.

## [2.3.2]
  Fixed issue where the backwards search would incorrectly highlight text.

## [2.3.1]
  Fixed issue with text selection between symbols.

## [2.3.0]
  Added an option to have separate highlighting options for symbols and content.
  
## [2.2.4]
  If a symbol with no end symbol was included within a different symbol, nothing would get highlighted when trying to highlight from the scope of the symbols.
  Example: { [ | }
  Expected behavior: The cursor ("|") is placed within {}, therefore this should be highlighted, even though "[" is the closest symbol to the cursor

## [2.2.3]
- The hotkey actions will now take the length of the symbols into account. This means that custom symbols should now be
    handled properly.

## [2.2.2]
- Removed font size multiplicator, since it caused issues when zooming. 
    Reasoning: There is no official support through the TextEditorDecorationType class, and the font size cannot be dynamically determined through the vscode API, so this feature will not be maintained and was therefore removed.

## [2.2.1]
- Fixed issue where HighlightScopeFromText sometimes wouldn't work correctly

## [2.2.0]
- Added support for highlighting of symbols share the same opening or closing symbols
- Added option to change the font size of highlighted text

## [2.1.0]
- Added experimental support to ignore escaped symbols: To use it, enable the regex mode in the settings. If you find any errors, please open a ticket!
- Fixed uncaught error exceptions

## [2.0.0]
- Added hotkeys to jump to/out of highlighted symbols.
- Cursors which aren't in a highlighted range will be removed when jumping.
- Added hotkey to select all the text between symbols (not including the symbols).
- Made the default settings way less aggressive to look at.

## [1.8.1]
- Fixed bug where multi cursor highlighting wouldn't work.
- Fixed bug with the option "highlight scope from text" where the wrong symbols would be picked for highlighting if they were on the same line.

## [1.8.0]
- Added option to change the highlighted text color.

## [1.7.0]
- Added an option to only highlight the enclosing symbols. The text inside will not be highlighted.

## [1.6.0]
- Added support for highlighting with multiple cursors

## [1.5.0]
- Added a timeout for the highlighting action of the extension. This improves the performance when writing/erasing letters at the edge of a symbol, however the highlighting may be delayed in some unwanted cases (i.e. when just moving the cursor over symbols without writing). If there was no problem prior to this update, set the timeout value to 0 to get the old behavior. Lowering this value will make highlighting more responsive, however there might be a noticeable delay when writing/erasing.

## [1.4.3]
- Updated the store page to show all settings
- Made the example for using custom symbols more readable

## [1.4.2]
- Fixed a crash that occurred when the cursor was at position 0 of a line and there was no valid symbol to highlight.

## [1.4.1]
- Fixed a problem where symbols wouldn't always be found correctly to the left of the cursor

## [1.4.0]
- Added option to define custom opening and closing symbols (i.e. do ... end). For an example configuration look at the store page.
- Added option to disable the extension back to the settings
- Added hotkey to enable/disable the extension (Default: ctrl+alt+l)

## [1.3.1]
- Fixed a bug where text would wrongly be blurred out when following conditions applied:
    1. blurring text out enabled
    2. searching the scope from within the text enabled
    3. selecting a closing bracket and then searching the scope from within text

## [1.3.0]
- Feature request: Added option to highlight text when clicking anywhere inside of the scope (so it's not necessary to click on brackets directly). This will use the same enabled brackets as the normal process.

## [1.2.1] 
- Minor performance improvements when writing inside of a highlighted area

## [1.2.0]
- Reverted changes made in 1.0.9 and 1.1.0
- Users can now define a maximum of searched lines for a closing symbol. (Default: 1000) This way accidentally clicking on symbols enclosing an extremely big amount of lines will not be searched endlessly.

## [1.1.0]
- Changed the extension to only highlight text lines that are currently visible. That way the extension will not take ages/freeze the editor when highlighting huge files. If this causes trouble for anyone please use version 1.0.8 and tell me about the problems!

## [1.0.9]
- Added an icon for the extension. 

## [1.0.8]
- Changed default value of language identifiers to an empty field. (I.e. it is enabled globally)

## [1.0.7]
- Added support to explicitly only support configured languages. Once this setting has one or more entries, the extension will ONLY work for the given entry/entries.
For language identifiers please refer to https://code.visualstudio.com/docs/languages/identifiers
Leaving the option blank will enable it globally.

## [1.0.6]
- Fixed case where text that spans over exactly 2 lines wouldn't correctly be highlighted when searching forwards

## [1.0.5]
- Added option to blur out non-highlighted text (Thanks @VSCode Dimmer Block for the idea)

src/ConfigHandler.ts
import * as vscode from 'vscode';
import { DecorationType } from './DecorationHandler';
import DecorationOptions from './DecorationOptions';

export enum JumpBetweenStrategy {
    TO_SYMBOL_START = "toSymbolStart",
    TO_SYMBOL_OPPOSITE_SIDE = "toSymbolOppositeSide"
}

export class HighlightEntry {
    constructor(public readonly symbol: string, public readonly isRegex: boolean, public readonly canBeSubstring: boolean) {

    }
}

export class HighlightSymbol {
    constructor(public readonly startSymbol: HighlightEntry, public readonly endSymbol: HighlightEntry, public readonly jumpBetweenStrategy: JumpBetweenStrategy) { }
}

export default class ConfigHandler {
    constructor() { }

    private getConfiguration() {
        return vscode.workspace.getConfiguration("BracketHighlighter", null);
    }

    public blurOutOfScopeText(): boolean {
        const config = this.getConfiguration();
        let blurOutOfScopeText: boolean | undefined = config.get("blurOutOfScopeText");
        if (blurOutOfScopeText === undefined) {
            blurOutOfScopeText = false;
        }
        return blurOutOfScopeText;
    }

    public getOpacity(): string {
        const config = this.getConfiguration();
        let opacity: string | undefined = config.get("blurOpacity");
        if (opacity === undefined) {
            opacity = '0.5';
        }
        return opacity;
    }

    public activeWhenDebugging(): boolean {
        const config = this.getConfiguration();
        let activeInDebug: boolean | undefined = config.get("activeInDebugMode");
        if (activeInDebug === undefined) {
            activeInDebug = true;
        }
        return activeInDebug;
    }

    public getMaxLineSearchCount(): number {
        const config = this.getConfiguration();
        let maxLineSearchCount: number | undefined = config.get("maxLineSearchCount");
        if (maxLineSearchCount === undefined) {
            maxLineSearchCount = 1000;
        }
        return maxLineSearchCount;
    }

    public highlightScopeFromText(): boolean {
        const config = this.getConfiguration();
        let highlightScopeFromText: boolean | undefined = config.get("highlightScopeFromText");
        if (highlightScopeFromText === undefined) {
            highlightScopeFromText = false;
        }
        return highlightScopeFromText;
    }

    public isInsideOfOpeningSymbolIgnored(): boolean {
        const config = this.getConfiguration();
        let isIgnored: boolean | undefined = config.get("isInsideOfOpeningSymbolIgnored");
        if (isIgnored === undefined) {
            isIgnored = false;
        }
        return isIgnored
    }

    public isInsideOfClosingSymbolIgnored(): boolean {
        const config = this.getConfiguration();
        let isIgnored: boolean | undefined = config.get("isInsideOfClosingSymbolIgnored");
        if (isIgnored === undefined) {
            isIgnored = false;
        }
        return isIgnored
    }

    public getDecorationOptions(decorationType: DecorationType): DecorationOptions {
        const config = this.getConfiguration();
        let fontWeight: string | undefined = undefined;
        let fontStyle: string | undefined = undefined;
        let letterSpacing: string | undefined = undefined;
        let outline: string | undefined = undefined;
        let border: string | undefined = undefined;
        let backgroundColor: string | undefined = undefined;
        let textDecoration: string | undefined = undefined;
        let textColor: string | undefined = undefined;
        if (decorationType === DecorationType.SYMBOLS && this.isDifferentSymbolHighlightingUsed()) {
            fontWeight = config.get("fontWeightSymbol");
            fontStyle = config.get("fontStyleSymbol");
            letterSpacing = config.get("letterSpacingSymbol");
            outline = config.get("outlineSymbol");
            border = config.get("borderSymbol");
            backgroundColor = config.get("backgroundColorSymbol");
            textDecoration = config.get("textDecorationSymbol");
            textColor = config.get("textColorSymbol");
        }
        else {
            fontWeight = config.get("fontWeight");
            fontStyle = config.get("fontStyle");
            letterSpacing = config.get("letterSpacing");
            outline = config.get("outline");
            border = config.get("border");
            backgroundColor = config.get("backgroundColor");
            textDecoration = config.get("textDecoration");
            textColor = config.get("textColor");
        }
        return new DecorationOptions(fontWeight, fontStyle, letterSpacing, outline, border, textDecoration, backgroundColor, textColor);
    }

    private isDifferentSymbolHighlightingUsed(): boolean {
        const config = this.getConfiguration();
        let differentHighlightingUsed: boolean | undefined = config.get("differentSymbolHighlightingUsed");
        if (!differentHighlightingUsed) {
            differentHighlightingUsed = false;
        }
        return differentHighlightingUsed;
    }

    public getEnabledLanguages(): Array<string> {
        const config = this.getConfiguration();
        let allowedLanguageIdString: string | undefined = config.get("allowedLanguageIds");
        if (allowedLanguageIdString === undefined) {
            return [];
        }
        allowedLanguageIdString = allowedLanguageIdString.replace(/\s/g, "");
        let allowedLanguageIds: Array<string> = allowedLanguageIdString.split(",");
        return allowedLanguageIds;
    }

    private getAcceptedSymbols(configuredSymbols: any[]): HighlightSymbol[] {
        let acceptedSymbols: HighlightSymbol[] = [];
        for (let customSymbol of configuredSymbols) {
            let strategy = customSymbol.hasOwnProperty("jumpBetweenStrategy") ? customSymbol.jumpBetweenStrategy : this.getDefaultJumpBetweenStrategy();
            /* New symbol syntax */
            if (customSymbol.hasOwnProperty("highlightPair")) {
                let entries: HighlightEntry[] = [];
                for (let pair of customSymbol.highlightPair) {
                    if (pair.hasOwnProperty("symbol")) {
                        let isRegex = pair.hasOwnProperty("isRegex") ? pair.isRegex : false;
                        let canBeSubstring = pair.hasOwnProperty("canBeSubstring") ? pair.canBeSubstring : false;
                        if (isRegex) {
                            canBeSubstring = false;
                        }

                        entries.push(new HighlightEntry(pair.symbol, isRegex, canBeSubstring));
                    }
                }
                if (entries.length === 2) {
                    acceptedSymbols.push(new HighlightSymbol(entries[0], entries[1], strategy));
                }
            }
            /* Support old symbol syntax */
            else if (customSymbol.hasOwnProperty("open") && customSymbol.hasOwnProperty("close") &&
                customSymbol.open !== customSymbol.close) {
                let entries = [new HighlightEntry(customSymbol.open, false, false), new HighlightEntry(customSymbol.close, false, false)];
                acceptedSymbols.push(new HighlightSymbol(entries[0], entries[1], strategy));
            }
        }
        return acceptedSymbols;
    }

    public getConfiguredSymbols(): HighlightSymbol[] {
        const config = this.getConfiguration();
        let configuredSymbols: any[] | undefined = config.get("customSymbols");
        if (configuredSymbols === undefined) {
            configuredSymbols = [];
        }
        return this.getAcceptedSymbols(configuredSymbols);
    }

    public isExtensionEnabled(): boolean {
        const config = this.getConfiguration();
        let extensionEnabled: boolean | undefined = config.get("enableExtension");
        if (extensionEnabled === undefined) {
            extensionEnabled = true;
        }
        return extensionEnabled;
    }

    public setExtensionEnabledStatus(extensionEnabled: boolean) {
        let config = this.getConfiguration();
        config.update("enableExtension", extensionEnabled);
    }

    public getTimeOutValue(): number {
        let config = this.getConfiguration();
        let timeOutValue: number | undefined = config.get("timeOutValue");
        if (timeOutValue === undefined) {
            timeOutValue = 600;
        }
        return timeOutValue;
    }

    public ignoreContent(): boolean {
        let config = this.getConfiguration();
        let ignoreContent: boolean | undefined = config.get("ignoreContent");
        if (ignoreContent === undefined) {
            ignoreContent = true;
        }
        return ignoreContent;
    }

    public getTextColor(): string {
        const config = this.getConfiguration();
        let textColor: string | undefined = config.get("textColor");
        if (textColor === undefined) {
            textColor = '';
        }
        return textColor;
    }

    public getDefaultJumpBetweenStrategy(): JumpBetweenStrategy {
        let config = this.getConfiguration();
        let strategy: JumpBetweenStrategy;
        let strategyStr: string | undefined = config.get("defaultJumpBetweenStrategy");
        let defaultStrategy: JumpBetweenStrategy = JumpBetweenStrategy.TO_SYMBOL_START;
        if (strategyStr === undefined) {
            strategy = defaultStrategy;
        }
        else {
            const allStrategies = Object.values(JumpBetweenStrategy);
            strategy = strategyStr as JumpBetweenStrategy;
            const isValidStrategy = allStrategies.indexOf(strategy) >= 0;
            if (!isValidStrategy) {
                strategy = defaultStrategy;
            }
        }
        return strategy;
    }

    public preferredJumpBetweenStrategiesBySymbol(): Map<string, JumpBetweenStrategy> {
        const config = this.getConfiguration();
        const map = new Map<string, JumpBetweenStrategy>();
        const customSymbols: any = config.get("customSymbols");
        if (Array.isArray(customSymbols)) {
            const allStrategies = Object.values(JumpBetweenStrategy);
            for (let customSymbol of customSymbols) {
                const isGoodSymbolPairs =
                    customSymbol.hasOwnProperty("open") &&
                    customSymbol.hasOwnProperty("close") &&
                    customSymbol.open !== customSymbol.close;
                if (isGoodSymbolPairs) {
                    if (customSymbol.hasOwnProperty("preferredJumpBetweenStrategy")) {
                        const strategyStr = customSymbol.preferredJumpBetweenStrategy;
                        const strategy = strategyStr as JumpBetweenStrategy;
                        const isValidStrategy = allStrategies.indexOf(strategy) >= 0;
                        if (isValidStrategy) {
                            // Let's be lazy and use the open as the key. I think it's okay for now.
                            map.set(customSymbol.open, strategy);
                        }
                    }
                }
            }
        }
        return map;
    }
}

export { ConfigHandler };